File: js/base/websocket/websocket_base_connection.js
"use strict";

const EventEmitter = require('events')
const {
    NotSupported
} = require('../errors')

module.exports = class WebsocketBaseConnection extends EventEmitter {
    /*
    Events: 
        on('open')
        on('close')
        on('err', reason)
        on('message', data)
    */
    constructor () {
        super();
    }

    connect () {
        throw new NotSupported('not implemented method <connect>');
    }

    close () {
        throw new NotSupported('not implemented method <close>');
    }

    send(data) {
        throw new NotSupported('not implemented method <send>');
    }

    isActive() {
        throw new NotSupported('not implemented method <isActive>');
    }

    sendJson(data) {
        this.send (JSON.stringify(data));
    }
};

File: python/ccxt/async_support/websocket/websocket_base_connection.py
# -*- coding: utf-8 -*-

""" AsyncConnection """

__all__ = [
    'WebsocketBaseConnection'
]

from pyee import EventEmitter
from abc import ABC, abstractmethod
import json


class WebsocketBaseConnection (ABC, EventEmitter):
    def __init__(self):
        super(WebsocketBaseConnection, self).__init__()

    @abstractmethod
    def connect(self):
        pass

    @abstractmethod
    def close(self):
        pass

    @abstractmethod
    def send(self, data):
        pass

    def sendJson(self, data):
        self.send(json.dumps(data))
